namerd/storage/zk/src/main/scala/com/twitter/finagle/serverset2/client/buoyant/ZkSession.scala

```diff
@@ -123,71 +124,83 @@ class ZkSession(
       @volatile var closed = false
 
       def loop(): Future[Unit] = {
-        if (!closed) safeRetry(go).respond {
+        if (closed) Future.Unit
+        else safeRetry(go).transform {
           case Throw(e@KeeperException.SessionExpired(_)) =>
             // don't retry. The session has expired while trying to set the watch.
             // In case our activity is still active, notify the listener
             u() = Activity.Failed(e)
+            Future.Unit
 
           case Throw(exc) =>
             logger.error("Operation failed with %s. Session %s", exc, sessionId)
             u() = Activity.Failed(exc)
-            fireAndForget { retryWithDelay { loop() } }
+            retryWithDelay { loop() }
 
           case Return(Watched(value, state)) =>
             val ok = Activity.Ok(value)
             retryBackoff.reset()
             u() = ok
 
-            val _ = state.changes.respond {
-              case WatchState.Pending =>
-              // Ignore updates WatchState is Pending.
+            state.map { s =>
```

Comment:
This could use a comment explaining that this map is just to perform side effects (logging and updating the activity)

Reply:
I like that right now, the result of each operation is changed together with the addition of comments for each stage. I think that introducing an intermediate variable could possibly reduce readability as a reader may get hung up on the naming of that variable. I can separate the chain if you feel strongly about it, but otherwise I inclined to leave it. Let me know your thoughts!